<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>References · PEtab.jl</title><meta name="title" content="References · PEtab.jl"/><meta property="og:title" content="References · PEtab.jl"/><meta property="twitter:title" content="References · PEtab.jl"/><meta name="description" content="Documentation for PEtab.jl."/><meta property="og:description" content="Documentation for PEtab.jl."/><meta property="twitter:description" content="Documentation for PEtab.jl."/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script><link href="../assets/custom_theme.css" rel="stylesheet" type="text/css"/></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">PEtab.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../tutorial/">Tutorial</a></li><li><input class="collapse-toggle" id="menuitem-3" type="checkbox"/><label class="tocitem" for="menuitem-3"><span class="docs-label">Extended Tutorials</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../petab_simcond/">Simulation conditions</a></li><li><a class="tocitem" href="../petab_preeq_simulations/">Steady-State Simulations (Pre-Equilibration)</a></li><li><a class="tocitem" href="../petab_event/">Events (callbacks, dosages, etc.)</a></li><li><a class="tocitem" href="../petab_cond_specific/">Simulation Condition-Specific Parameters</a></li><li><a class="tocitem" href="../petab_obs_noise/">Noise and Observable Parameters</a></li><li><a class="tocitem" href="../import_petab/">Importing PEtab Standard Format</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-4" type="checkbox"/><label class="tocitem" for="menuitem-4"><span class="docs-label">Parameter Estimation</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../pest_method/">Parameter Estimation Methods</a></li><li><a class="tocitem" href="../pest_plot/">Plotting Estimation Results</a></li><li><a class="tocitem" href="../pest_algs/">Available and Recommended Algorithms</a></li><li><a class="tocitem" href="../pest_select/">Model Selection with PEtab-select</a></li><li><a class="tocitem" href="../pest_custom/">Wrapping Optimization Packages</a></li></ul></li><li><a class="tocitem" href="../inference/">Bayesian Inference</a></li><li><a class="tocitem" href="../API/">API</a></li><li><a class="tocitem" href="../grad_hess_methods/">Gradient and Hessian Methods</a></li><li><a class="tocitem" href="../default_options/">Default Options</a></li><li><input class="collapse-toggle" id="menuitem-9" type="checkbox"/><label class="tocitem" for="menuitem-9"><span class="docs-label">Performance Tips</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../nonstiff_models/">Non-Biology (Non-Stiff) Models</a></li><li><a class="tocitem" href="../Beer/">Condition-Specific Parameters</a></li><li><a class="tocitem" href="../Bachmann/">Adjoint Sensitivity Analysis (Large Models)</a></li></ul></li><li class="is-active"><a class="tocitem" href>References</a></li><li><a class="tocitem" href="../FAQ/">FAQ</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>References</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>References</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/sebapersson/PEtab.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/sebapersson/PEtab.jl/blob/main/docs/src/references.md#" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="References"><a class="docs-heading-anchor" href="#References">References</a><a id="References-1"></a><a class="docs-heading-anchor-permalink" href="#References" title="Permalink"></a></h1><p>Throughout the documentation, references related to parameter estimation for ODE models can be found. This page contains a complete list of all references mentioned in the documentation.</p><div class="citation canonical"><dl><dt>[1]</dt><dd><div id="frohlich2022fides">F. Fröhlich and P. K. Sorger. <em>Fides: Reliable trust-region optimization for parameter estimation of ordinary differential equation models</em>. PLoS computational biology <strong>18</strong>, e1010322 (2022).</div></dd><dt>[2]</dt><dd><div id="raue2013lessons">A. Raue, M. Schilling, J. Bachmann, A. Matteson, M. Schelke, D. Kaschek, S. Hug, C. Kreutz, B. D. Harms, F. J. Theis and others. <em>Lessons learned from quantitative dynamical modeling in systems biology</em>. PloS one <strong>8</strong>, e74335 (2013).</div></dd><dt>[3]</dt><dd><div id="hass2019benchmark">H. Hass, C. Loos, E. Raimundez-Alvarez, J. Timmer, J. Hasenauer and C. Kreutz. <em>Benchmark problems for dynamic modeling of intracellular processes</em>. Bioinformatics <strong>35</strong>, 3073–3082 (2019).</div></dd><dt>[4]</dt><dd><div id="wickham2014tidy">H. Wickham. <em>Tidy data</em>. Journal of statistical software <strong>59</strong>, 1–23 (2014).</div></dd><dt>[5]</dt><dd><div id="villaverde2019benchmarking">A. F. Villaverde, F. Fröhlich, D. Weindl, J. Hasenauer and J. R. Banga. <em>Benchmarking optimization methods for parameter estimation in large kinetic models</em>. Bioinformatics <strong>35</strong>, 830–838 (2019).</div></dd><dt>[6]</dt><dd><div id="frohlich2017parameter">F. Fröhlich, F. J. Theis, J. O. Rädler and J. Hasenauer. <em>Parameter estimation for dynamical systems with discrete events and logical operations</em>. Bioinformatics <strong>33</strong>, 1049–1056 (2017).</div></dd><dt>[7]</dt><dd><div id="schmiester2021petab">L. Schmiester, Y. Schälte, F. T. Bergmann, T. Camba, E. Dudkin, J. Egert, F. Fröhlich, L. Fuhrmann, A. L. Hauber, S. Kemmer and others. <em>PEtab—Interoperable specification of parameter estimation problems in systems biology</em>. PLoS computational biology <strong>17</strong>, e1008646 (2021).</div></dd><dt>[8]</dt><dd><div id="boehm2014identification">M. E. Boehm, L. Adlung, M. Schilling, S. Roth, U. Klingmüller and W. D. Lehmann. <em>Identification of isoform-specific dynamics in phosphorylation-dependent STAT5 dimerization by quantitative mass spectrometry and mathematical modeling</em>. Journal of proteome research <strong>13</strong>, 5685–5694 (2014).</div></dd><dt>[9]</dt><dd><div id="villaverde2022protocol">A. F. Villaverde, D. Pathirana, F. Fröhlich, J. Hasenauer and J. R. Banga. <em>A protocol for dynamic model calibration</em>. Briefings in bioinformatics <strong>23</strong>, bbab387 (2022).</div></dd><dt>[10]</dt><dd><div id="wachter2006implementation">A. Wächter and L. T. Biegler. <em>On the implementation of an interior-point filter line-search algorithm for large-scale nonlinear programming</em>. Mathematical programming <strong>106</strong>, 25–57 (2006).</div></dd><dt>[11]</dt><dd><div id="gabel2019famos">M. Gabel, T. Hohl, A. Imle, O. T. Fackler and F. Graw. <em>FAMoS: A Flexible and dynamic Algorithm for Model Selection to analyse complex systems dynamics</em>. PLOS Computational Biology <strong>15</strong>, e1007230 (2019).</div></dd><dt>[12]</dt><dd><div id="vihola2014ergonomic">M. Vihola. <em>Ergonomic and reliable Bayesian inference with adaptive Markov chain Monte Carlo</em>. Wiley statsRef: statistics reference online, 1–12 (2014).</div></dd><dt>[13]</dt><dd><div id="hoffman2014no">M. D. Hoffman, A. Gelman and others. <em>The No-U-Turn sampler: adaptively setting path lengths in Hamiltonian Monte Carlo.</em> J. Mach. Learn. Res. <strong>15</strong>, 1593–1623 (2014).</div></dd><dt>[14]</dt><dd><div id="carpenter2017stan">B. Carpenter, A. Gelman, M. D. Hoffman, D. Lee, B. Goodrich, M. Betancourt, M. A. Brubaker, J. Guo, P. Li and A. Riddell. <em>Stan: A probabilistic programming language</em>. Journal of statistical software <strong>76</strong> (2017).</div></dd><dt>[15]</dt><dd><div id="gelman2020bayesian">A. Gelman, A. Vehtari, D. Simpson, C. C. Margossian, B. Carpenter, Y. Yao, L. Kennedy, J. Gabry, P.-C. Bürkner and M. Modrák. <em>Bayesian workflow</em>, arXiv preprint arXiv:2011.01808 (2020).</div></dd><dt>[16]</dt><dd><div id="sapienza2024differentiable">F. Sapienza, J. Bolibar, F. Schäfer, B. Groenke, A. Pal, V. Boussange, P. Heimbach, G. Hooker, F. Pérez, P.-O. Persson and others. <em>Differentiable Programming for Differential Equations: A Review</em>, arXiv preprint arXiv:2406.09699 (2024).</div></dd><dt>[17]</dt><dd><div id="blondel2024elements">M. Blondel and V. Roulet. <em>The elements of differentiable programming</em>, arXiv preprint arXiv:2403.14606 (2024).</div></dd><dt>[18]</dt><dd><div id="revels2016forward">J. Revels, M. Lubin and T. Papamarkou. <em>Forward-mode automatic differentiation in Julia</em>, arXiv preprint arXiv:1607.07892 (2016).</div></dd><dt>[19]</dt><dd><div id="mester2022differential">R. Mester, A. Landeros, C. Rackauckas and K. Lange. <em>Differential methods for assessing sensitivity in biological models</em>. PLoS computational biology <strong>18</strong>, e1009598 (2022).</div></dd><dt>[20]</dt><dd><div id="frohlich2017scalable">F. Fröhlich, B. Kaltenbacher, F. J. Theis and J. Hasenauer. <em>Scalable parameter estimation for genome-scale biochemical reaction networks</em>. PLoS computational biology <strong>13</strong>, e1005331 (2017).</div></dd><dt>[21]</dt><dd><div id="ma2021comparison">Y. Ma, V. Dixit, M. J. Innes, X. Guo and C. Rackauckas. <em>A comparison of automatic differentiation and continuous sensitivity analysis for derivatives of differential equation solutions</em>. In: <em>2021 IEEE High Performance Extreme Computing Conference (HPEC)</em> (IEEE, 2021); pp. 1–9.</div></dd><dt>[22]</dt><dd><div id="raue2015data2dynamics">A. Raue, B. Steiert, M. Schelker, C. Kreutz, T. Maiwald, H. Hass, J. Vanlier, C. Tönsing, L. Adlung, R. Engesser and others. <em>Data2Dynamics: a modeling environment tailored to parameter estimation in dynamical systems</em>. Bioinformatics <strong>31</strong>, 3558–3560 (2015).</div></dd><dt>[23]</dt><dd><div id="stadter2021benchmarking">P. Städter, Y. Schälte, L. Schmiester, J. Hasenauer and P. L. Stapor. <em>Benchmarking of numerical integration methods for ODE models of biological systems</em>. Scientific reports <strong>11</strong>, 2696 (2021).</div></dd><dt>[24]</dt><dd><div id="beer2014creating">R. Beer, K. Herbst, N. Ignatiadis, I. Kats, L. Adlung, H. Meyer, D. Niopek, T. Christiansen, F. Georgi, N. Kurzawa and others. <em>Creating functional engineered variants of the single-module non-ribosomal peptide synthetase IndC by T domain exchange</em>. Molecular BioSystems <strong>10</strong>, 1709–1718 (2014).</div></dd><dt>[25]</dt><dd><div id="bachmann2011division">J. Bachmann, A. Raue, M. Schilling, M. E. Böhm, C. Kreutz, D. Kaschek, H. Busch, N. Gretz, W. D. Lehmann, J. Timmer and others. <em>Division of labor by dual feedback regulators controls JAK2/STAT5 signaling over broad ligand range</em>. Molecular systems biology <strong>7</strong>, 516 (2011).</div></dd></dl></div></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../Bachmann/">« Adjoint Sensitivity Analysis (Large Models)</a><a class="docs-footer-nextpage" href="../FAQ/">FAQ »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.13.0 on <span class="colophon-date" title="Monday 7 July 2025 19:59">Monday 7 July 2025</span>. Using Julia version 1.11.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
