import{_ as a,c as i,o as n,aA as r}from"./chunks/framework.BJSyVkCt.js";const g=JSON.parse('{"title":"References","description":"","frontmatter":{},"headers":[],"relativePath":"references.md","filePath":"references.md","lastUpdated":null}'),o={name:"references.md"};function t(s,e,l,m,d,c){return n(),i("div",null,[...e[0]||(e[0]=[r('<h1 id="References" tabindex="-1">References <a class="header-anchor" href="#References" aria-label="Permalink to &quot;References {#References}&quot;">​</a></h1><p>Throughout the documentation, references related to parameter estimation for ODE models can be found. This page contains a complete list of all references mentioned in the documentation.</p><ol><li><p><a id="persson2025petab"></a> S. Persson, F. Fröhlich, S. Grein, T. Loman, D. Ognissanti, V. Hasselgren, J. Hasenauer and M. Cvijovic. <em>PEtab. jl: advancing the efficiency and utility of dynamic modelling</em>. Bioinformatics <strong>41</strong>, btaf497 (2025).</p></li><li><p><a id="frohlich2022fides"></a> F. Fröhlich and P. K. Sorger. <em>Fides: Reliable trust-region optimization for parameter estimation of ordinary differential equation models</em>. PLoS computational biology <strong>18</strong>, e1010322 (2022).</p></li><li><p><a id="raue2013lessons"></a> A. Raue, M. Schilling, J. Bachmann, A. Matteson, M. Schelke, D. Kaschek, S. Hug, C. Kreutz, B. D. Harms, F. J. Theis and others. <em>Lessons learned from quantitative dynamical modeling in systems biology</em>. PloS one <strong>8</strong>, e74335 (2013).</p></li><li><p><a id="hass2019benchmark"></a> H. Hass, C. Loos, E. Raimundez-Alvarez, J. Timmer, J. Hasenauer and C. Kreutz. <em>Benchmark problems for dynamic modeling of intracellular processes</em>. Bioinformatics <strong>35</strong>, 3073–3082 (2019).</p></li><li><p><a id="wickham2014tidy"></a> H. Wickham. <em>Tidy data</em>. Journal of statistical software <strong>59</strong>, 1–23 (2014).</p></li><li><p><a id="villaverde2019benchmarking"></a> A. F. Villaverde, F. Fröhlich, D. Weindl, J. Hasenauer and J. R. Banga. <em>Benchmarking optimization methods for parameter estimation in large kinetic models</em>. Bioinformatics <strong>35</strong>, 830–838 (2019).</p></li><li><p><a id="frohlich2017parameter"></a> F. Fröhlich, F. J. Theis, J. O. Rädler and J. Hasenauer. <em>Parameter estimation for dynamical systems with discrete events and logical operations</em>. Bioinformatics <strong>33</strong>, 1049–1056 (2017).</p></li><li><p><a id="schmiester2021petab"></a> L. Schmiester, Y. Schälte, F. T. Bergmann, T. Camba, E. Dudkin, J. Egert, F. Fröhlich, L. Fuhrmann, A. L. Hauber, S. Kemmer and others. <em>PEtab—Interoperable specification of parameter estimation problems in systems biology</em>. PLoS computational biology <strong>17</strong>, e1008646 (2021).</p></li><li><p><a id="jost2025petab"></a> P. J. Jost, F. T. Bergmann, D. Weindl and J. Hasenauer. <em>PEtab-GUI: A graphical user interface to create, edit and inspect PEtab parameter estimation problems</em>, arXiv preprint arXiv:2511.11515 (2025).</p></li><li><p><a id="boehm2014identification"></a> M. E. Boehm, L. Adlung, M. Schilling, S. Roth, U. Klingmüller and W. D. Lehmann. <em>Identification of isoform-specific dynamics in phosphorylation-dependent STAT5 dimerization by quantitative mass spectrometry and mathematical modeling</em>. Journal of proteome research <strong>13</strong>, 5685–5694 (2014).</p></li><li><p><a id="pathirana2025petab"></a> D. Pathirana, F. T. Bergmann, D. Doresic, P. Lakrisenko, S. Persson, N. Neubrand, J. Timmer, C. Kreutz, H. Binder, M. Cvijovic and others. <em>PEtab Select: specification standard and supporting software for automated model selection</em>, bioRxiv, 2025–05 (2025).</p></li><li><p><a id="villaverde2022protocol"></a> A. F. Villaverde, D. Pathirana, F. Fröhlich, J. Hasenauer and J. R. Banga. <em>A protocol for dynamic model calibration</em>. Briefings in bioinformatics <strong>23</strong>, bbab387 (2022).</p></li><li><p><a id="wachter2006implementation"></a> A. Wächter and L. T. Biegler. <em>On the implementation of an interior-point filter line-search algorithm for large-scale nonlinear programming</em>. Mathematical programming <strong>106</strong>, 25–57 (2006).</p></li><li><p><a id="gabel2019famos"></a> M. Gabel, T. Hohl, A. Imle, O. T. Fackler and F. Graw. <em>FAMoS: A Flexible and dynamic Algorithm for Model Selection to analyse complex systems dynamics</em>. PLOS Computational Biology <strong>15</strong>, e1007230 (2019).</p></li><li><p><a id="vihola2014ergonomic"></a> M. Vihola. <em>Ergonomic and reliable Bayesian inference with adaptive Markov chain Monte Carlo</em>. Wiley statsRef: statistics reference online, 1–12 (2014).</p></li><li><p><a id="hoffman2014no"></a> M. D. Hoffman, A. Gelman and others. <em>The No-U-Turn sampler: adaptively setting path lengths in Hamiltonian Monte Carlo.</em> J. Mach. Learn. Res. <strong>15</strong>, 1593–1623 (2014).</p></li><li><p><a id="carpenter2017stan"></a> B. Carpenter, A. Gelman, M. D. Hoffman, D. Lee, B. Goodrich, M. Betancourt, M. A. Brubaker, J. Guo, P. Li and A. Riddell. <em>Stan: A probabilistic programming language</em>. Journal of statistical software <strong>76</strong> (2017).</p></li><li><p><a id="gelman2020bayesian"></a> A. Gelman, A. Vehtari, D. Simpson, C. C. Margossian, B. Carpenter, Y. Yao, L. Kennedy, J. Gabry, P.-C. Bürkner and M. Modrák. <em>Bayesian workflow</em>, arXiv preprint arXiv:2011.01808 (2020).</p></li><li><p><a id="stadter2021benchmarking"></a> P. Städter, Y. Schälte, L. Schmiester, J. Hasenauer and P. L. Stapor. <em>Benchmarking of numerical integration methods for ODE models of biological systems</em>. Scientific reports <strong>11</strong>, 2696 (2021).</p></li><li><p><a id="sapienza2024differentiable"></a> F. Sapienza, J. Bolibar, F. Schäfer, B. Groenke, A. Pal, V. Boussange, P. Heimbach, G. Hooker, F. Pérez, P.-O. Persson and others. <em>Differentiable Programming for Differential Equations: A Review</em>, arXiv preprint arXiv:2406.09699 (2024).</p></li><li><p><a id="blondel2024elements"></a> M. Blondel and V. Roulet. <em>The elements of differentiable programming</em>, arXiv preprint arXiv:2403.14606 (2024).</p></li><li><p><a id="revels2016forward"></a> J. Revels, M. Lubin and T. Papamarkou. <em>Forward-mode automatic differentiation in Julia</em>, arXiv preprint arXiv:1607.07892 (2016).</p></li><li><p><a id="mester2022differential"></a> R. Mester, A. Landeros, C. Rackauckas and K. Lange. <em>Differential methods for assessing sensitivity in biological models</em>. PLoS computational biology <strong>18</strong>, e1009598 (2022).</p></li><li><p><a id="frohlich2017scalable"></a> F. Fröhlich, B. Kaltenbacher, F. J. Theis and J. Hasenauer. <em>Scalable parameter estimation for genome-scale biochemical reaction networks</em>. PLoS computational biology <strong>13</strong>, e1005331 (2017).</p></li><li><p><a id="ma2021comparison"></a> Y. Ma, V. Dixit, M. J. Innes, X. Guo and C. Rackauckas. <em>A comparison of automatic differentiation and continuous sensitivity analysis for derivatives of differential equation solutions</em>. In: <em>2021 IEEE High Performance Extreme Computing Conference (HPEC)</em> (IEEE, 2021); pp. 1–9.</p></li><li><p><a id="raue2015data2dynamics"></a> A. Raue, B. Steiert, M. Schelker, C. Kreutz, T. Maiwald, H. Hass, J. Vanlier, C. Tönsing, L. Adlung, R. Engesser and others. <em>Data2Dynamics: a modeling environment tailored to parameter estimation in dynamical systems</em>. Bioinformatics <strong>31</strong>, 3558–3560 (2015).</p></li><li><p><a id="beer2014creating"></a> R. Beer, K. Herbst, N. Ignatiadis, I. Kats, L. Adlung, H. Meyer, D. Niopek, T. Christiansen, F. Georgi, N. Kurzawa and others. <em>Creating functional engineered variants of the single-module non-ribosomal peptide synthetase IndC by T domain exchange</em>. Molecular BioSystems <strong>10</strong>, 1709–1718 (2014).</p></li><li><p><a id="bachmann2011division"></a> J. Bachmann, A. Raue, M. Schilling, M. E. Böhm, C. Kreutz, D. Kaschek, H. Busch, N. Gretz, W. D. Lehmann, J. Timmer and others. <em>Division of labor by dual feedback regulators controls JAK2/STAT5 signaling over broad ligand range</em>. Molecular systems biology <strong>7</strong>, 516 (2011).</p></li></ol>',3)])])}const f=a(o,[["render",t]]);export{g as __pageData,f as default};
